docs/RingBuffer.md

```diff
@@ -0,0 +1,312 @@
+ * Consume available ring buffer(s) data without event polling.
+ * Returns number of records consumed across all registered ring buffers (or
+ * INT_MAX, whichever is less), or negative number if any of the callbacks
+ * return error.
+ *
+ * @param[in] rb Pointer to ring buffer manager.
+ */
+int ring_buffer__consume(struct ring_buffer *rb);
+
+/**
+ * @brief Get a handle that can be used to sleep until data is available in the ring(s).
+ *
+ * @param[in] rb Pointer to ring buffer manager.
+ *
+ * @return ebpf_invalid_fd on error
+ * @return fd_t corresponding to I/O Completion port hand on success
+ */
+fd_t ring_buffer__epoll_fd(const struct ring_buffer *rb);
+```
+
+### New ebpf APIs
+
+```c
+/**
+ * get pointers to mapped producer and consumer pages
+ *
+ * @param[out] producer pointer* to start of read-only mapped producer pages
+ * @param[out] consumer pointer* to start of read-write mapped consumer page
+ */
+int ebpf_ring_buffer_get_buffer(void **producer, void **consumer);
+```
+
+## Ringbuffer consumer
+
+### libbpf mapped memory consumer example
+
+
+This consumer directly accesses the records from the producer memory and directly updates the consumer offset to show the logic.
+
+```c++
+
+//
+// == Ringbuf helpers ==
+//
+
+// Ring buffer record is 32bit header + data.
+typedef struct _rb_header
+{
+    //NOTE: bit fields are not portable, so this is just for simpler example code -- the actual code should use bit masking to perform equivalent operations on the header bits.
+    uint8_t locked : 1;
+    uint8_t discarded : 1;
+    uint32_t length : 30;
+} rb_header_t;
+
+typedef struct _rb_record
+{
+    rb_header_t header;
+    uint8_t data[];
+} rb_record_t;
+
+/**
+ * @brief clear the ringbuffer.
+ */
+void rb_flush(uint64_t *cons_offset, const uint64_t *prod_offset) {
+    *cons_offset = *prod_offset;
+}
+
+
+//
+// == mmap/epoll consumer ==
+//
+
+void *rb_cons; // Pointer to read/write mapped consumer page with consumer offset.
+void *rb_prod; // Pointer to start of read-only producer pages.
+
+// Open ringbuffer.
+fd_t map_fd = bpf_obj_get(rb_map_name.c_str());
+if (map_fd == ebpf_fd_invalid) return 1;
+
+auto rb = ring_buffer__new(
+    map_fd,
+    NULL, //callback function for callback-based consumer
+    nullptr, nullptr);
+if (!rb) return 1;
+
+// get pointers to the producer/consumer pages
+int err = ebpf_ring_buffer_get_buffer(&rb_prod, &rb_cons, /* &epoll_fd */);
+
+if (err) {
+    goto Cleanup;
+}
+
+const uint64_t *prod_offset = (const uint64_t*)rb_prod; // Producer offset ptr (r only).
+uint64_t *cons_offset = (uint64_t*)rb_cons; // Consumer offset ptr (r/w mapped).
+const uint8_t *rb_data = ((const uint8_t*)rb_prod) + PAGESIZE; // Double-mapped rb data ptr (r only).
+
+// have_data used to track whether we should wait for notification or just keep reading.
+bool have_data = *prod_offset > *cons_offset;
+
+// Initialize iocp wait handle.
+epoll_fd = ring_buffer__epoll_fd(rb);
```

Comment:
Should this come before line 193?

Reply:
It doesn't have to but that probably makes more sense. The code (that was) on line 194 gets points to the mapped memory, and could be an alternative way to get a waitable handle (instead of using epoll_fd to get an integer that can be converted to a HANDLE).

I am modifying this based on discussions, so the next version won't be calling `epoll_fd()` at all (just `poll()`), so we can match the libbpf API without needing extra wrappers for epoll.